lsp/roslyn_ls.lua

```diff
@@ -0,0 +1,195 @@
+---@param target string
+local function on_init_sln(client, target)
+  vim.notify("Initializing: " .. target, vim.log.levels.INFO, { title = "roslyn-ls" })
+  ---@diagnostic disable-next-line: param-type-mismatch
+  client:request("solution/open", {
+      solution = vim.uri_from_fname(target),
+  },
+  function (err, result, context)
+  end)
+end
+
+---@param client vim.lsp.Client
+---@param project_files string[]
+local function on_init_project(client, project_files)
+  vim.notify("Initializing: projects", vim.log.levels.INFO, { title = "roslyn-ls" })
+  ---@diagnostic disable-next-line: param-type-mismatch
+  client:request("project/open", {
+      projects = vim.tbl_map(function(file)
+          return vim.uri_from_fname(file)
+      end, project_files),
+  },
+  function (err, result, context)
+  end)
+end
+
+local function roslyn_handlers()
+  return {
+        ["workspace/projectInitializationComplete"] = function(_, _, ctx)
+            vim.notify("Roslyn project initialization complete", vim.log.levels.INFO, { title = "roslyn.nvim" })
+
+            local buffers = vim.lsp.get_buffers_by_client_id(ctx.client_id)
+            for _, buf in ipairs(buffers) do
+                vim.lsp.util._refresh("textDocument/diagnostic", { bufnr = buf })
+            end
+
+            ---NOTE: This is used by rzls.nvim for init
+            vim.api.nvim_exec_autocmds("User", { pattern = "RoslynInitialized", modeline = false })
+        end,
+        ["workspace/_roslyn_projectHasUnresolvedDependencies"] = function()
+            vim.notify("Detected missing dependencies. Run dotnet restore command.", vim.log.levels.ERROR, {
+                title = "roslyn.nvim",
+            })
+            return vim.NIL
+        end,
+        ["workspace/_roslyn_projectNeedsRestore"] = function(_, result, ctx)
+            local client = assert(vim.lsp.get_client_by_id(ctx.client_id))
+
+            ---@diagnostic disable-next-line: param-type-mismatch
+            client:request("workspace/_roslyn_restore", result, function(err, response)
+                if err then
+                    vim.notify(err.message, vim.log.levels.ERROR, { title = "roslyn.nvim" })
+                end
+                if response then
+                    for _, v in ipairs(response) do
+                        vim.notify(v.message, vim.log.levels.INFO, { title = "roslyn.nvim" })
+                    end
+                end
+            end)
+
+            return vim.NIL
+        end,
+        ["razor/provideDynamicFileInfo"] = function(_, _, _)
+            vim.notify(
+                "Razor is not supported.\nPlease use https://github.com/tris203/rzls.nvim",
+                vim.log.levels.WARN,
+                { title = "roslyn.nvim" }
+            )
+            return vim.NIL
+        end,
+    }
+end
+
+---@type vim.lsp.ClientConfig
+return {
+  name = "roslyn_ls",
+  offset_encoding = 'utf-8',
+  ---@param dispatchers vim.lsp.rpc.Dispatchers
+  cmd = function (dispatchers)
```

Comment:
IMO, `cmd` should not be set by default, and should instead be set explicitly by the user. I think an example should be using `--stdio` flag for starting up the server

Reply:
I do think it would make sense to have a default command and using `--stdio` here make sense, as it's simpler to handle here.

Said that, it's tricky here to define a proper default `cmd`, because there is no common executable name that can be used here, I need to have another look on how the server is distributed, as there are platform specific nuget packages and some which are platform independent (which contain only `.dll` files that have to be called with `dotnet` command).

Reply:
> Said that, it's tricky here to define a proper default cmd, because there is no common executable name that can be used here, I need to have another look on how the server is distributed, as there are platform specific nuget packages and some which are platform independent (which contain only .dll files that have to be called with dotnet command).

Exactly, that is why I don't think it makes sense to have a default cmd value, and instead have docs on an example

Reply:
I personally had issues to setup `omnisharp` back then because there was no default `cmd` and having something that works for some people and can be used as an example to write your own `cmd` is IMO better then having only something documented which should only theoretically work, but this is only my very personal opinion.

I just checked and in the platform specific nuget packages there are `Microsoft.CodeAnalysis.LanguageServer.exe` and `Microsoft.CodeAnalysis.LanguageServer` executables in the `/content/LanguageServer` dir. So this could be used instead, so if the user put it in the path it can be picked up properly.

Try to test this also and adjust here after that.

Reply:
Yeah I sort of agree with it being cumbersome to setup without a default cmd, but when there is bad support in package managers for this server (I haven't seen a single one have it), it is also annoying having to add things to path

I don't have any big objections to this if others think it should be there. There at least needs to be documentation on how to install it and how to make the default cmd work

Reply:
This is where I hope we can merge https://github.com/williamboman/mason.nvim/pull/1781 in future, I use this currently in combination with this package definition https://github.com/616b2f/mason-registry/blob/main/packages/roslyn/package.yaml.

Other alternative is to compile from source, or download and extract nuget files manually, none of which is convenient IMO.

I will try to write something up in the documentation part